### PR TITLE
fix: dedupe external import binding for format `esm`

### LINKED ISSUE #1548: Same named export imported by different files result in unnecessary as aliases
### Reproduction link or steps

To reproduce, run `pnpm build vue -r` in https://github.com/vuejs/core/tree/rolldown (make sure to checkout `rolldown` branch)

Then check `packages/vue/dist/vue.esm-browser.js` line 2

### What is expected?

```js
import { parse, parseExpression } from "@babel/parser";
```

### What is actually happening?

The same `parseExpression` import from `@babel/parser` is aliased two times:

```js
import { parse, parseExpression, parseExpression as parseExpression$1, parseExpression as parseExpression$2 } from "@babel/parser";
```

### System Info

```Shell
NA
```

### Any additional comments?

_No response_

### PR DESCRIPTION
**Pull request by IWANABETHATGUY**
### Description
1. fixed https://github.com/rolldown/rolldown/issues/1548

### CODE REVIEW

File: crates/rolldown/src/stages/link_stage/bind_imports_and_exports.rs [hunk 1]
```diff
@@ -355,13 +366,17 @@ struct BindImportsAndExportsContext<'a> {
...
+    FxHashMap<ModuleIdx, FxHashMap<CompactStr, IndexSet<SymbolRef>>>,
 }
 
 impl BindImportsAndExportsContext<'_> {
+  #[allow(clippy::too_many_lines)]
   fn match_imports_with_exports(&mut self, module_id: ModuleIdx) {
     let Module::Normal(module) = &self.index_modules[module_id] else {
       return;
     };
+    let is_esm = matches!(self.options.format, OutputFormat::Esm);
```
Comment (IWANABETHATGUY) [inline crates/rolldown/src/stages/link_stage/bind_imports_and_exports.rs:379]: Only merge `import` binding in esm format, for other format it is not necessary.

File: crates/rolldown/tests/esbuild/default/forbid_string_export_names_no_bundle/artifacts.snap [hunk 2]
```diff
@@ -8,11 +8,11 @@ snapshot_kind: text
...
 import * as name space from "./foo";
-import { name 1 as name 2, same name } from "./foo";
+import { name 1, same name } from "./foo";
 
 //#region entry.js
 let ok = true;
 
 //#endregion
-export { name 2, name space, ok as 'not ok', ok, same name };
+export { name 1 as 'name 2', name space, ok as 'not ok', ok, same name };
```
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/default/forbid_string_export_names_no_bundle/artifacts.snap:17]: not related to this pr.

File: crates/rolldown/tests/esbuild/default/forbid_string_import_names_bundle/artifacts.snap [hunk 3]
```diff
@@ -20,7 +21,7 @@ source: crates/rolldown_testing/src/integration_test.rs
 ## entry.js
 
 ```js
-import { some import as nested } from "external";
+import { some import } from "external";
 
-export { nested };
+export { some import as nested };
```
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/default/forbid_string_import_names_bundle/artifacts.snap:26]: not related to this pr.

File: crates/rolldown/tests/esbuild/default/forbid_string_import_names_no_bundle/artifacts.snap [hunk 4]
```diff
@@ -7,11 +7,11 @@ snapshot_kind: text
...
-import { an import as anImport, another import as an export } from "./foo";
+import { an import, another import } from "./foo";
 
 //#region entry.js
-anImport();
+an import();
 
 //#endregion
-export { an export };
+export { another import as 'an export' };
```
Comment (IWANABETHATGUY) [inline crates/rolldown/tests/esbuild/default/forbid_string_import_names_no_bundle/artifacts.snap:16]: As you can see the output is already wrong before, I will fix it in another pr.

Comment (IWANABETHATGUY) [reply crates/rolldown/tests/esbuild/default/forbid_string_import_names_no_bundle/artifacts.snap:16]: https://github.com/rolldown/rolldown/issues/3094

File: crates/rolldown/src/stages/link_stage/bind_imports_and_exports.rs (hunk 1 above)
Comment (IWANABETHATGUY) [reply crates/rolldown/src/stages/link_stage/bind_imports_and_exports.rs:379]: because the namedImport is already merged for other format

**Review by hyf0 — APPROVED** [review]

**Pull request merged**